sys/net/credman/credman.c

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (C) 2019 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     net_credman
+ * @{
+ *
+ * @file
+ * @brief       (D)TLS Credentials management module implementation
+ *
+ * @author  Aiman Ismail <muhammadaimanbin.ismail@haw-hamburg.de>
+ */
+
+#include "net/credman.h"
+#include "mutex.h"
+
+#include <string.h>
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+static mutex_t _mutex = MUTEX_INIT;
+
+static credman_credential_t credentials[CREDMAN_MAX_CREDENTIALS];
+static unsigned used = 0;
+
+static int _find_credential_pos(credman_tag_t tag, credman_type_t type);
+
+int credman_add(const credman_credential_t *credential)
+{
+    assert(credential);
+    mutex_lock(&_mutex);
+    int pos = -1;
+    int ret = CREDMAN_ERROR;
+
+    if ((credential->type == CREDMAN_TYPE_EMPTY) ||
+        (credential->tag == CREDMAN_TAG_EMPTY)) {
+        DEBUG("credman: invalid credential type/tag\n");
+        ret = CREDMAN_INVALID;
+        goto end;
+    }
+    switch (credential->type) {
+    case CREDMAN_TYPE_PSK:
+        if ((credential->params.psk.key.s == NULL) ||
+            (credential->params.psk.key.len == 0)) {
+            DEBUG("credman: invalid PSK parameters\n");
+            ret = CREDMAN_INVALID;
+            goto end;
+        }
+        break;
+    case CREDMAN_TYPE_ECDSA:
+        if ((credential->params.ecdsa.private_key == NULL) ||
+            (credential->params.ecdsa.public_key.x == NULL) ||
+            (credential->params.ecdsa.public_key.y == NULL)) {
+            DEBUG("credman: invalid ECDSA parameters\n");
+            ret = CREDMAN_INVALID;
+            goto end;
+        }
+        break;
+    default:
+        ret = CREDMAN_TYPE_UNKNOWN;
+        goto end;
+    }
+    pos = _find_credential_pos(credential->tag, credential->type);
+    if (pos >= 0) {
+        DEBUG("credman: credential with tag %d and type %d already exist\n",
+              credential->tag, credential->type);
+        ret = CREDMAN_EXIST;
+        goto end;
+    }
+    /* find the next free position in credential pool */
+    pos = _find_credential_pos(CREDMAN_TAG_EMPTY, CREDMAN_TYPE_EMPTY);
+    if (pos < 0) {
+        DEBUG("credman: no space for new credential\n");
+        ret = CREDMAN_NO_SPACE;
+        goto end;
+    }
```

Comment:
Though I'm wondering if this

```C
pos = _find_credential_pos(tag, type);
if (check pos condition) {
    throw error;
}
```

duplication might be consolidatable somehow. I had no idea yet though.

Reply:
Usually @cgundogan has some nice ideas when it comes to this kind of stuff. Maybe you can ask him :-).

Reply:
This is what I came up with based on suggestion from Cenk:

```
/* Find position of credential with given tag and type.
 * To avoid going through the list twice, this function also returns the
 * first empty position found.
 */
static int _find_credential_pos(credman_tag_t tag, credman_type_t type, int8_t *empty_pos)
{
    *empty_pos = -1;
    for (unsigned i = 0; i < CREDMAN_MAX_CREDENTIALS; i++) {
        credman_credential_t *c = &credentials[i];
        if ((c->tag == tag) && (c->type == type)) {
            return i;
        }
        /* only check until empty position found */
        if ((empty_pos) && (*empty_pos < 0) &&
            (c->tag == CREDMAN_TAG_EMPTY) && (c->type == CREDMAN_TYPE_EMPTY)) {
            *empty_pos = i;
        }
    }
    return -1;
}

...

/* check if credential already exist and find empty position simultaneously */
int8_t empty_pos;
    pos = _find_credential_pos(credential->tag, credential->type, &empty_pos);
    if (pos >= 0) {
        DEBUG("credman: credential with tag %d and type %d already exist\n",
              credential->tag, credential->type);
        ret = CREDMAN_EXIST;
        goto end;
    }
    if (empty_pos < 0) {
        DEBUG("credman: no space for new credential\n");
        ret = CREDMAN_NO_SPACE;
        goto end;
    }
```

Reply:
I can't find your comment on this in Github, but I found the following in my mails

> This is what I came up with based on suggestion from Cenk:
> ```C
> /* Find position of credential with given tag and type.
>  * To avoid going through the list twice, this function also returns the
>  * first empty position found.
>  */
> static int _find_credential_pos(credman_tag_t tag, credman_type_t type, int8_t *empty_pos)
> {
>     *empty_pos = -1;
>     for (unsigned i = 0; i < CREDMAN_MAX_CREDENTIALS; i++) {
>         credman_credential_t *c = &credentials[i];
>         if ((c->tag == tag) && (c->type == type)) {
>             return i;
>         }
>         /* only check until empty position found */
>         if ((empty_pos) && (*empty_pos < 0) &&
>             (c->tag == CREDMAN_TAG_EMPTY) && (c->type == CREDMAN_TYPE_EMPTY)) {
>             *empty_pos = i;
>         }
>     }
>     return -1;
> }
> 
> ...
> 
> /* check if credential already exist and find empty position simultaneously */
> int8_t empty_pos;
>     pos = _find_credential_pos(credential->tag, credential->type, &empty_pos);
>     if (pos >= 0) {
>         DEBUG("credman: credential with tag %d and type %d already exist\n",
>               credential->tag, credential->type);
>         ret = CREDMAN_EXIST;
>         goto end;
>     }
>     if (empty_pos < 0) {
>         DEBUG("credman: no space for new credential\n");
>         ret = CREDMAN_NO_SPACE;
>         goto end;
>     }
> ```

How about the following simplification (also note, that I was able to do all without `goto`; I kept the label for the error handling above though):

```C
/* Find position of credential with given tag and type.
 * To avoid going through the list twice, this function also returns the
 * first empty credential pool entry found.
 */
static int _find_credential_pos(credman_tag_t tag, credman_type_t type,
credman_credential_t **empty)
{
    for (unsigned i = 0; i < CREDMAN_MAX_CREDENTIALS; i++) {
        credman_credential_t *c = &credentials[i];
        if ((c->tag == tag) && (c->type == type)) {
            return i;
        }
        /* only check until empty position found */
        if ((empty) && (*empty != NULL) &&
            (c->tag == CREDMAN_TAG_EMPTY) && (c->type == CREDMAN_TYPE_EMPTY)) {
            *empty = c;
        }
    }
    return -1;
}

...

int credman_add(const credman_credential_t *credential)
{
    credman_credential_t *entry = NULL;
    assert(credential);
...
    pos = _find_credential_pos(credential->tag, credential->type, &entry);
    if (pos >= 0) {
        DEBUG("credman: credential with tag %d and type %d already exist\n",
              credential->tag, credential->type);
        ret = CREDMAN_EXIST;
    }
    else if (entry == NULL) {
        DEBUG("credman: no space for new credential\n");
        ret = CREDMAN_NO_SPACE;
    }
    else {
        *entry = *credential;
        used++;
        ret = CREDMAN_OK;
    }
end:
    mutex_unlock(&_mutex);
    return ret;
}

Reply:
> I can't find your comment on this in Github,

I accidentally deleted my comment while editing it. Sorry.

The code looks better and simpler now even with no `goto`. I'll keep this pattern in mind for next time. I'll push a commit for this later today.